### PR TITLE
NIFI-10243: allow ControlRate to throttle on combination of data rate or flowfile rate

### PR DESCRIPTION
**Pull request by markobean**
# Summary
This PR modifies ControlRate processor to introduce a new mode of operation. Rate throttling can now be done by both Data Rate and FlowFile Rate. The first rate to be reached begins throttling. Documentation was improved and unit testing expanded to cover additional cases. Backward compatibility is maintained.

[NIFI-10243](https://issues.apache.org/jira/browse/NIFI-10243)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [X] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [X] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [X] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [X] Pull Request based on current revision of the `main` branch
- [X] Pull Request refers to a feature branch with one commit containing changes

# Verification

Unit tests expanded to cover additional cases - both new modes of operation and older modes that were not adequately tested.

Instantiated processor in a running NiFi to perform real world testing.

Viewed documentation including new Additional Details page to verify proper representation.

### Build

- [X] Build completed using `mvn clean install -P contrib-check`
  - [X] JDK 8
  - [X] JDK 11
  - [X] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [X] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java [hunk 1]
```diff
@@ -211,16 +241,37 @@ public void onPropertyModified(final PropertyDescriptor descriptor, final String
                 || descriptor.equals(GROUPING_ATTRIBUTE_NAME)
                 || descriptor.equals(TIME_PERIOD)) {
             // if the criteria that is being used to determine limits/throttles is changed, we must clear our throttle map.
-            throttleMap.clear();
-        } else if (descriptor.equals(MAX_RATE)) {
+            dataThrottleMap.clear();
+            countThrottleMap.clear();
+        } else if (descriptor.equals(MAX_RATE) || descriptor.equals(MAX_DATA_RATE)) {
+            // MAX_RATE could affect eitehr throttle map; MAX_DATA_RATE only affects data throttle map
```
Comment (michael81877) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java:247]:
```suggestion
            // MAX_RATE could affect either throttle map; MAX_DATA_RATE only affects data throttle map
```

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/resources/docs/org.apache.nifi.processors.standard.ControlRate/additionalDetails.html [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                <td>FlowFiles are counted regardless of content size. No more than the specified number of FlowFiles pass through this processor in the given Time Duration. The Maximum Rate property
+                    specifies the maximum number of FlowFiles allowed per Time Duration.</td>
+            </tr>
+            <tr>
+                <td>Attribute Value</td>
+                <td>The value of an attribute is accumulated to determine overall rate. The Rate Controlled Attribute property specifies the attribute whose value will be accumulated. The value of
+                    the specified attribute is expected to be an integer. This mode is independent of overall FlowFile size and count.</td>
+            </tr>
+            <tr>
+                <td>Data Rate or FLowFile Count</td>
```
Comment (michael81877) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/resources/docs/org.apache.nifi.processors.standard.ControlRate/additionalDetails.html:51]:
```suggestion
                <td>Data Rate or FlowFile Count</td>
```

**Comment by thenatog** [comment]
Will review

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestControlRate.java [hunk 3]
```diff
@@ -258,10 +282,245 @@ public void testNonExistingGroupAttribute() throws InterruptedException {
...
+
+        // Increase rate after stopping processor. Previous count should remain since we are still inside time period
+        runner.setProperty(ControlRate.MAX_RATE, "33 B");
+        runner.run(7, false);
+        runner.assertTransferCount(ControlRate.REL_SUCCESS, 3);
+        runner.assertTransferCount(ControlRate.REL_FAILURE, 0);
+        runner.assertQueueNotEmpty();
+
+        // after 1 second, we should be able to send the up to 3 more flowfiles
+        Thread.sleep(ONE_SEC_PLUS);
```
Comment (thenatog) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestControlRate.java:313]: I'm wary of adding tests which used thread sleeps. They are frequently unreliable when run on Github Actions infrastructure. The obvious problem is that the it's hard to test a control rate feature without using a delay like this.. Anyway, thought it was worth mentioning, as we're trying to reduce the number of flakey tests we have.

Comment (markobean) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestControlRate.java:313]: Agree. I don't care for thread sleeps either. I'm open to suggestions if you have any. Still, the existing unit tests uses sleeps and so my addition is consistent. And, if you noticed, I reduced the total sleep time (prior to adding new unit tests) by reducing the amount of time spent sleeping. Personally, that's an issue I see frequently - sleeping much longer than necessary which has a cumulative affect on the overall build. In this case, a minimum time of 1 sec is required due to the minimum time allowed by the property. And, the sleep is only 1.01 sec.

In other words, I feel it's the best that can be done given the situation and the component being tested.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java [hunk 4]
```diff
@@ -71,30 +70,57 @@ public class ControlRate extends AbstractProcessor {
...
+            .description("The maximum rate at which FlowFiles should pass through this processor. The format of this property is expected to be a "
+                    + "positive integer representing FlowFiles count per Time Duration")
+            .required(false)
+            .addValidator(StandardValidators.POSITIVE_LONG_VALIDATOR)
+            .dependsOn(RATE_CONTROL_CRITERIA, DATA_OR_FLOWFILE_RATE)
+            .build();
+
     public static final PropertyDescriptor RATE_CONTROL_ATTRIBUTE_NAME = new PropertyDescriptor.Builder()
             .name("Rate Controlled Attribute")
             .description("The name of an attribute whose values build toward the rate limit if Rate Control Criteria is set to 'attribute value'. "
```
Comment (thenatog) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java:126]: Should the RATE_CONTROL_ATTRIBUTE_NAME descriptor also dependsOn 'ATTRIBUTE_RATE_VALUE' or should it always be available to configure?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java [hunk 5]
```diff
@@ -71,30 +70,57 @@ public class ControlRate extends AbstractProcessor {
     public static final String DATA_RATE = "data rate";
```
Comment (thenatog) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java:70]: Small issue but I would like to see these capitalized (Data Rate) rather than all lower case as is currently the case. Not sure if this will impact existing flows - if that's the case, leave it as is.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java [hunk 6]
```diff
@@ -148,6 +176,8 @@ protected void init(final ProcessorInitializationContext context) {
         final List<PropertyDescriptor> properties = new ArrayList<>();
         properties.add(RATE_CONTROL_CRITERIA);
         properties.add(MAX_RATE);
+        properties.add(MAX_DATA_RATE);
+        properties.add(MAX_COUNT_RATE);
         properties.add(RATE_CONTROL_ATTRIBUTE_NAME);
         properties.add(TIME_PERIOD);
```
Comment (thenatog) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java:182]: The 'Time Period' descriptor should probably be moved to second or top position in the order, as it's a mandatory field and also the most relevant for the processor

**Review by thenatog — COMMENTED** [review]
I tested the ControlRate processor manually and found each of the Rate Controls to operate as expected. I have left a few comments and I'm still reviewing the code to see what else could be improved if anything

**Comment by thenatog** [comment]
I briefly reviewed the code changes and everything seems to make sense. I think ideally the inner classes which already existed could be moved to external files to clear out the main class somewhat - potentially something for another Jira task.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java (hunk 4 above)
Comment (markobean) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java:126]: Good call. The `dependsOn` wasn't an option when this property was first created. It makes sense to add it now (even though not explicitly part of this ticket.)

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestControlRate.java (hunk 3 above)
Comment (thenatog) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestControlRate.java:313]: Is there a way to reduce the time to less than a second? I see that the TIME_PERIOD has a validator for 1 second, but potentially if we could override this and other spots that may expect a 1 second period, to make it shorter overall for testing. Build time with tests on my personal machine is already pretty long, and build time on Github actions is frequently timing out, so we're trying to reduce as much of this in tests as possible. @exceptionfactory may have some advice.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java (hunk 5 above)
Comment (markobean) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java:70]: I believe this is a breaking change that would affect existing flows. At least, when I updated the values to have a different case, e.g. "Data Rate", all unit tests failed. I don't see a big deal with this, and to avoid potential problems I think it should be left as-is.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java (hunk 6 above)
Comment (markobean) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java:182]: Moved to 2nd from top.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestControlRate.java (hunk 3 above)
Comment (markobean) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestControlRate.java:313]: I really like the idea of reducing build times. However, I'm not aware of how to override a validator - nor do I believe that is a good idea. Such an override seems like it may introduce false positive results inadvertently. 

I would prefer to take up the issue of reducing build times in a separate JIRA ticket as it is not directly related to the feature proposed here.

**Comment by markobean** [comment]
I think refactoring the inner classes is scope creep. Also, those classes are scoped as private implying there is specific intent to have them as inner classes. I recommend leaving them as-is or opening a separate ticket for further evaluation and work.

Thanks for the review @thenatog. I believe all your comments have been resolved. Is this ready to be merged?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java [hunk 7]
```diff
@@ -268,48 +336,67 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
...
-                rateValue = Long.parseLong(attributeValue);
-                break;
-            default:
-                throw new AssertionError("<Rate Control Criteria> property set to illegal value of " + rateControlCriteria);
+    /*
+     * Determine the amount this FlowFile will incur against the maximum allowed rate.
+     * This is applicable to counting accruals, flowfiles or attributes
+     */
+    private long getCountAccrual(FlowFile flowFile) {
+        long rateValue = -1L;
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java:374]: It would be helpful to define a `private static final` value for the default value of `-1` and reuse that in multiple places.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java [hunk 8]
```diff
@@ -408,34 +498,59 @@ public FlowFileFilterResult filter(FlowFile flowFile) {
...
+                    dataThrottle.setMaxRate(DataUnit.parseDataSize(maximumRateStr, DataUnit.B).longValue());
+                    dataThrottleMap.put(groupName, dataThrottle);
                 }
-                throttle.setMaxRate(newRate);
 
-                throttleMap.put(groupName, throttle);
+                dataThrottle.lock();
+                try {
+                    if (dataThrottle.tryAdd(getDataSizeAccrual(flowFile))) {
+                        flowFilesInBatch += 1;
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java:515]:
```suggestion
                        flowFilesInBatch++;
```

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java [hunk 9]
```diff
@@ -408,34 +498,59 @@ public FlowFileFilterResult filter(FlowFile flowFile) {
...
+            if (countThrottleRequired() && !dataThrottlingActive) {
+                if (countThrottle == null) {
+                    countThrottle = new Throttle(timePeriodSeconds, TimeUnit.SECONDS, getLogger());
+                    countThrottle.setMaxRate(Long.parseLong(maximumCountRateStr));
+                    countThrottleMap.put(groupName, countThrottle);
+                }
+                countThrottle.lock();
+                try {
+                    if (countThrottle.tryAdd(getCountAccrual(flowFile))) {
+                        flowFilesInBatch += 1;
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java:543]:
```suggestion
                        flowFilesInBatch++;
```

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for working on this new feature @markobean!

The general runtime implementation looks good, the main concern I see is the additional unit test time given the one second sleep. As an interim solution, setting the time period to `500 ms` and and the sleep time to `510` would help. A follow-on issue could incorporate adjustments to the Processor that would allow overriding a method that provides the current time. For this PR, simply reducing the sleep time would be helpful.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java (hunk 7 above)
Comment (markobean) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java]: Done.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java (hunk 8 above)
Comment (markobean) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java]: Done.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java (hunk 9 above)
Comment (markobean) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ControlRate.java]: Done.

**Comment by markobean** [comment]
@exceptionfactory The time period cannot be reduced to 500 ms. The validator requires a minimum value of 1 sec. This is because the exact rate becomes less accurate the smaller the time period, especially once in sub-second range (and made worse on a busy system.) In order to change the unit tests to a smaller value, the validator would have to change. I do not think that is an advisable approach - to allow configurations which could mislead users in terms of accuracy for the sake of shortening unit tests. In order to test properly, the sleep time is a necessary evil.

**Comment by exceptionfactory** [comment]
Thanks for making the updates and noting the detail about the minimum validation setting @markobean, the changes look good.

The new test method push the total test time to over 12 seconds for standard execution, which will place it among the slower unit tests, contributing to an overall slow build process. I have an idea for updating the unit test to extend ControlRate for the test class and implement an alternative current time method. With that background, I'm not opposed to moving forward with this PR, and I can evaluate an improvement in a subsequent Jira issue.

I will defer to @thenatog for final review at this point, thanks!

**Comment by thenatog** [comment]
+1 will merge, maybe in future we can find some test efficiencies if they became a problem.
